src/Compiler/Checking/NameResolution.fs

```diff
@@ -2725,6 +2725,33 @@ let rec ResolveLongIdentInTypePrim (ncenv: NameResolver) nenv lookupKind (resInf
         let errorTextF s =
             match tryTcrefOfAppTy g ty with
             | ValueSome tcref ->
+                if tcref.IsRecordTycon then
+                    let alternative = nenv.eFieldLabels |> Map.tryFind nm
+                    match alternative with
+                    | Some fieldLabels ->
+                        let fieldsOfResolvedType = tcref.AllFieldsArray  |> Array.map (fun f -> f.LogicalName) |> Set.ofArray
+                        let fieldsOfAlternatives =
+                            fieldLabels
+                            |> Seq.collect (fun l -> l.Tycon.AllFieldsArray |> Array.map (fun f -> f.LogicalName))
+                            |> Set.ofSeq
+                        let intersect = Set.intersect fieldsOfAlternatives fieldsOfResolvedType
+
+                        if not intersect.IsEmpty then
+                            let resolvedTypeName = $"""{System.String.Join(".", tcref.CompilationPath.DemangledPath)}.{tcref.LogicalName}"""
```

Comment:
No, at least I didn't find one and looking around in the codebase, it's often done like this.
But yes, that doesn't feel all that great.

Reply:
Things like should go via memoization, to prevent string allocations many times.
the NicePrint module should be using something at least similar to inspire from.

Reply:
I changed it to use `NicePrint.layoutTyconRefImpl`.